### PR TITLE
ASN Range models

### LINKED ISSUE #111: Add support for allocating a pool of BGP ASNs
### Environment
* Nautobot version:  1.5.14
* nautobot-bgp-models version:  0.7.1

### Proposed Functionality
The functionality would be to add a feature to add a range of BGP ASNs to a Tenant or Location. 
 
### Use Case
The user is currently assigning 1000s of ASNs to each data center using a separate application. They would like to be able to document this pool of ASNs in Nautobot and map it to the appropriate data center.

### LINKED ISSUE #117: ASN Grouping
### Environment
* Nautobot version: 1.5.0
* nautobot-bgp-models version:  0.7.1

### Proposed Functionality
**ASN Grouping.** Nautobot should be able to group, nest, or something similar ASNs. Nautobot should treat this similar to prefixes with containers.

### Use Case
This allows a team to associate a range of ASNs to a region, site, team, organization, etc.

### PR DESCRIPTION
**Pull request by mzbroch**
This PR introduces ASN Range model.

Closes #111 
Closes #117

### CODE REVIEW

**Comment by mzbroch** [comment]
Hey @glennmatthews  , can I ask for early review of this proposal ? Thanks !

File: nautobot_bgp_models/views.py [hunk 1]
```diff
@@ -27,6 +29,70 @@ class AutonomousSystemUIViewSet(NautobotUIViewSet):
...
+        """Return any additional context data for the template."""
+        context = super().get_extra_context(request, instance)
+        if self.action == "retrieve":
+
+            def add_available_asns(instance, asns):
+                """Create fake records for all gaps between used Autonomous Systems."""
+                if not asns:
+                    return [{"asn": instance.asn_min, "available": instance.asn_max - instance.asn_min + 1}]
+
+                prev_asn = instance.asn_max
```
Comment (glennmatthews) [inline nautobot_bgp_models/views.py:54]: Should this be `instance.asn_min`? I'll admit I'm having a bit of trouble following the logic here, some comments would be appreciated.

**Review by glennmatthews — COMMENTED** [review]
Looks like a reasonable starting point for this model.

File: nautobot_bgp_models/models.py [hunk 2]
```diff
@@ -143,6 +144,49 @@ def __str__(self):
...
+        verbose_name = "Autonomous System Range"
+
+    def __str__(self):
+        """String representation of an AutonomousSystemRange."""
+        return f"ASN Range {self.asn_min}-{self.asn_max}"
+
+    def clean(self):
+        """Clean."""
+        if self.asn_min >= self.asn_max:
+            raise ValidationError("asn_min must be lower than asn_max.")
```
[the remark this answers is not in the document]
Comment (mzbroch) [reply nautobot_bgp_models/models.py:176]: Good point , however I have not seen this as a requirement 👍

File: nautobot_bgp_models/models.py [hunk 3]
```diff
@@ -143,6 +144,49 @@ def __str__(self):
...
+    "webhooks",
+)
+class AutonomousSystemRange(PrimaryModel):
+    """Autonomous System Range information."""
+
+    name = models.CharField(max_length=100, unique=True, blank=False)
+    asn_min = ASNField(verbose_name="Start", help_text="Min value for 32-bit autonomous system number")
+    asn_max = ASNField(verbose_name="End", help_text="Max value for 32-bit autonomous system number")
+    description = models.CharField(max_length=200, blank=True)
+    tenant = models.ForeignKey(to=Tenant, on_delete=models.PROTECT, blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (mzbroch) [reply nautobot_bgp_models/models.py:163]: It makes sense to introduce `tenant` on `AutonomousSystem` in separate PR 👍 , would You agree ?

File: nautobot_bgp_models/navigation.py [hunk 4]
```diff
@@ -26,6 +26,21 @@
...
+                        permissions=["nautobot_bgp_models.view_autonomoussystemrange"],
+                        buttons=(
+                            NavMenuAddButton(
+                                link="plugins:nautobot_bgp_models:autonomoussystemrange_add",
+                                permissions=["nautobot_bgp_models.add_autonomoussystemrange"],
+                            ),
+                            NavMenuImportButton(
+                                link="plugins:nautobot_bgp_models:autonomoussystemrange_import",
+                                permissions=["nautobot_bgp_models.add_autonomoussystemrange"],
+                            ),
```
[the remark this answers is not in the document]
Comment (mzbroch) [reply nautobot_bgp_models/navigation.py:41]: Will address through #189

File: nautobot_bgp_models/views.py (hunk 1 above)
Comment (mzbroch) [reply nautobot_bgp_models/views.py]: Changed to new logic with unittests 👍

File: nautobot_bgp_models/filters.py [hunk 5]
```diff
@@ -40,6 +40,32 @@ class Meta:
...
+    CustomFieldModelFilterSetMixin,
+):
+    """Filtering of AutonomousSystemRange records."""
+
+    q = django_filters.CharFilter(
+        method="search",
+        label="Search",
+    )
+
+    def search(self, queryset, name, value):  # pylint: disable=unused-argument
```
Comment (glennmatthews) [inline nautobot_bgp_models/filters.py:55]: Would be better (and simpler) to use `nautobot.apps.filters.SearchFilter` for the `q` filter.

File: nautobot_bgp_models/models.py [hunk 6]
```diff
@@ -143,6 +144,49 @@ def __str__(self):
...
+        """String representation of an AutonomousSystemRange."""
+        return f"ASN Range {self.asn_min}-{self.asn_max}"
+
+    def clean(self):
+        """Clean."""
+        if self.asn_min >= self.asn_max:
+            raise ValidationError("asn_min value must be lower than asn_max value.")
+
+    def get_next_available_asn(self):
+        """Return the first available ASN number in the range."""
```
Comment (glennmatthews) [inline nautobot_bgp_models/models.py:179]:
```suggestion
        """Return the first available ASN number in the range, or None if none are available."""
```

File: nautobot_bgp_models/tests/test_views.py [hunk 7]
```diff
@@ -77,6 +77,58 @@ def setUpTestData(cls):
...
+    # # test_bulk_import_objects_with_permission = None
+    # # test_bulk_import_objects_with_constrained_permission = None
+    #
+    # def test_list_objects_with_permission(self):
+    #     super().test_list_objects_with_permission()
+    #
+    # def _get_base_url(self):
+    #     return "plugins:{}:{}_{{}}".format(  # pylint: disable=consider-using-f-string
+    #         self.model._meta.app_label, self.model._meta.model_name
+    #     )
```
Comment (glennmatthews) [inline nautobot_bgp_models/tests/test_views.py:95]: Remove?

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**